src/lib/http/http_server.pl

```diff
@@ -62,14 +61,44 @@
     call(Handler, Request, Response).
 
 call_handler(Path, Headers, Body, Handlers, Response) :-
-    chars_utf8bytes(CharBody, Body),
     member(post(Dcg, Handler), Handlers),
     phrase(Dcg, Path),!,
-    Request = http_request(Headers, CharBody),
+    build_response(Body, ResponseBody),
+    Request = http_request(Headers, ResponseBody),
     call(Handler, Request, Response).
 
 call_handler(_, _, _, http_response(404, "Not found")).
 
+build_response(ByteBody, text(TextBody)) :-
+    chars_utf8bytes(TextBody, ByteBody).
+build_response(ByteBody, binary(ByteBody)).
+
+send_response(Stream, http_response(StatusCode, text(TextResponse), Headers)) :-
+    format(Stream, "HTTP/1.0 ~d\r\n", [StatusCode]),
+    overwrite_header("Content-Type"-"text/plain", Headers, Headers0),
+    overwrite_header("Connection"-"Close", Headers0, Headers1),
+    write_headers(Stream, Headers1),
+    format(Stream, "\r\n~s", [TextResponse]).
+
+send_response(Stream, http_response(StatusCode, binary(BinaryResponse), Headers)) :-
+    format(Stream, "HTTP/1.0 ~d\r\n", [StatusCode]),
+    overwrite_header("Connection"-"Close", Headers, Headers0),
+    write_headers(Stream, Headers0),
+    format(Stream, "\r\n", []),
+    put_bytes(Stream, BinaryResponse).
+
+write_headers(Stream, Headers) :-
+    forall(member(Key-Value, Headers), format(Stream, "~s: ~s\r\n", [Key, Value])).
```

Comment:
Using `forall/2` is not good style, because the side-effects this produces cannot be tested easily (i.e., via Prolog queries).

A better approach in this example would be to write a DCG that declaratively *describes* the header, as a list of characters. You can use the non-terminal `format_//2` in the description .

When the header is completely described (`phrase(header(...), Cs)`), use a single `format/3` call to write the entire list of characters to the stream: `format(Stream, "~s", Cs)`.